### PR TITLE
CAFDI-7 - Create Multi TLD Gravity Model MVP

### PR DESCRIPTION
**Pull request by wsp-mbuckley**
Describe Changes
---
Add class to calibrate gravity model to multiple cost distributions.

- [ ] CAFDI-20 - Move gravity and jacobian functions to single_Core
- [ ] CAFDI-21 - implement multi_cost_distribution Class
- [ ] CAFDI-25 - implement MultiGravityModelResults
- [ ] CAFDI-22 - refactor _calibrate function
- [ ] CAFDI-23 - implement _gravity_function
- [ ] CAFDI-24 - implement _jacobian_function
- [ ] CAFDI-26 - Write multiTLD tests for run
- [ ] CAFDI-27 - Write tests for calibration
- [ ] CAFDI-28 - Write tests for error checking

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by BenTaylor-TfN** [comment]
Finished #6 now. Also updated the packaging system in #8.

Rebased this off of the new main to keep the commit messages in here a bit clearer. No significant changes that you weren't already away of.

File: src/caf/distribute/gravity_model/multi_area.py [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+
+    @property
+    def multi_achieved_band_shares(self) -> np.ndarray:
+        """Analogous to _achieved_band_shares but for a multi-tld"""
+        if self.achieved_cost_dist is None:
+            raise ValueError("Gravity model has not been run. achieved_band_share is not set.")
+        shares = []
+        for dist in self.achieved_cost_dist:
+            shares.append(dist.band_share_vals)
+        return np.concatenate(shares)
```
Comment (BenTaylor-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:236]: Would it be better for this function to just overload `achieved_band_share` from the base class? It looks like this is designed to work in the same way?

File: src/caf/distribute/gravity_model/multi_area.py [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+
+    def _create_seed_matrix(self, cost_distributions, cost_args, params_len):
+        base_mat = np.zeros(self.cost_matrix.shape)
```
Comment (BenTaylor-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:239]: Better off doing this to retain typing etc from base matrix
```suggestion
        base_mat = np.zeros_like(self.cost_matrix)
```

File: src/caf/distribute/gravity_model/multi_area.py [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+                    f"If this is expected, remove {cat} from your "
+                    "tlds dataframe before inputting."
+                )
+
+            distribution = MultiCostDistribution(
+                name=cat, cost_distribution=tld, zones=zones, function_params=self.init_params
+            )
+            dists.append(distribution)
+
+        return dists
```
Comment (BenTaylor-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:219]: If this function is entirely about translating some cost distributions that are in some odd format into one that is suitable for the cost distribution to use, does it not make more sense for the gravity model to take the output format of this function as it's expected input?

We can then add this function to a class that formats and stores the data?

File: src/caf/distribute/gravity_model/multi_area.py [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+                cost_convergence=self.achieved_convergence[dist.name],
+                value_distribution=self.achieved_distribution[dist.zones],
+                target_cost_distribution=dist.cost_distribution,
+                cost_function=self.cost_function,
+                cost_params=self._cost_params_to_kwargs(
+                    best_params[i * params_len : i * params_len + params_len]
+                ),
+            )
+
+            results[dist.name] = gresult
```
Comment (BenTaylor-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:339]: Slightly clearer
```suggestion
            result_i = GravityModelCalibrateResults(
                cost_distribution=self.achieved_cost_dist[i],
                cost_convergence=self.achieved_convergence[dist.name],
                value_distribution=self.achieved_distribution[dist.zones],
                target_cost_distribution=dist.cost_distribution,
                cost_function=self.cost_function,
                cost_params=self._cost_params_to_kwargs(
                    best_params[i * params_len : i * params_len + params_len]
                ),
            )

            results[dist.name] = result_i
```

**Review by BenTaylor-TfN — CHANGES_REQUESTED** [review]
Mostly fine. Added a few suggested changes for minor edits. Some bigger comments below.

Don't forget to fix linting errors etc.

### Imports
You need to be more consistent with your imports. They should be split into three sections, this makes it easier for a reader to see where different imports are coming from when reading the code later.
- Built ins - Any built in python libraries
- Third party - Anything that's being installed from PyPI / Conda 
- Local imports - Anything local to this package. This should also be wrapped in a pylint error disable message to prevent the linter error where it can't find the local packages.

Each of these sections should then be ordered, see the [style guide](https://google.github.io/styleguide/pyguide.html#313-imports-formatting) for full detail. It may be worth looking into [isort](https://pypi.org/project/isort/) to automate this.

I've added some suggestions for what these imports should look like for `singe_area.py`. I know they need updating in the following files, but please check elsewhere that you've changed imports:
- `core.py`
- `multi_area.py`

### Tests
Looking at your tests, this is probably a good time to start the conversation on testing standards, discussing the pros and cons of each of our techniques.

Initially, I'm going to use the following as a reference, taken from code in this review:
```Python
class TestUtils:
    def test_infill_costs(self, costs, infilled, expected_infilled):
        assert np.array_equal(np.round(expected_infilled, 3), np.round(infilled, 3))

    def test_tld_processing_length(self, distributions_sorted):
        assert len(distributions_sorted) == 4

    def test_tld_processing_type(self, distributions_sorted):
        assert isinstance(distributions_sorted[0], gm.MultiCostDistribution)
```
An initial guess tells me that these tests are for the `utils.py` module, and there is one test per function. This in itself, isn't an issue as we can update the tests to be more thorough later. However, I think the following principles should be followed:
- Each module of functionality should have its own module of tests
- Each function should have one (or more) classes encapsulating it's tests, with methods for: 
  - expected functionality
  - edge cases
  - warnings / error checks
- Tests should contain as much info as possible to demonstrate the expected behaviour, and fixtures are used simply as data setup / teardown.
  - The function being tested should at least be called in the test (and not in the fixture) to make it clear that's what is being tested.
  - This is more of a readability function that anything, why make it harder to read a test later than you have to?
  - In the case of the above, that means moving the following fixtures into the test body
    - `infilled`
    - `distributions_sorted`

File: src/caf/distribute/furness.py [hunk 5]
```diff
@@ -134,7 +135,7 @@ def doubly_constrained_furness(
             # Calculate the diff - leave early if met
             row_diff = (row_targets - np.sum(furnessed_mat, axis=1)) ** 2
             col_diff = (col_targets - np.sum(furnessed_mat, axis=0)) ** 2
-            cur_rmse = (np.sum(row_diff + col_diff) / n_vals) ** 0.5
+            cur_rmse = ((np.sum(row_diff) + np.sum(col_diff)) / n_vals) ** 0.5
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/furness.py:138]: This change means furnessing still works for none-square matrices. This didn't end up being used in what I wrote but may be in future.

File: src/caf/distribute/furness.py [hunk 6]
```diff
@@ -157,3 +158,84 @@ def doubly_constrained_furness(
...
+#         furnessed_mat *= diff_factor
+#
+#         row_diff = (row_targets - np.sum(furnessed_mat, axis=1)) ** 2
+#         col_diff = (col_targets - np.sum(furnessed_mat, axis=0)) ** 2
+#         cur_rmse = ((np.sum(row_diff) + np.sum(col_diff)) / n_vals) ** 0.5
+#         if cur_rmse < tol:
+#             print('debugging')
+#             early_exit = True
+#             break
+#     return furnessed_mat
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/furness.py:241]: In this branch yes, it's readded and fixed in the triple-furness branch

File: src/caf/distribute/gravity_model/__init__.py [hunk 7]
```diff
@@ -6,5 +6,10 @@
 
 # Models
 from caf.distribute.gravity_model.single_area import SingleAreaGravityModelCalibrator
+from caf.distribute.gravity_model.multi_area import (
+    MultiCostDistribution,
+    MultiAreaGravityModelCalibrator,
+    MultiDistInput,
+)
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/gravity_model/__init__.py:13]: What do you mean by this? single TLD only imports the calibrator

File: pyproject.toml [hunk 8]
```diff
@@ -101,7 +102,7 @@ add-ignore = [
 # Tests
 [tool.pytest.ini_options]
 pythonpath = ["src"]
-addopts = '--cov=caf --cov-report=xml -n auto'
+addopts = '--cov=caf --cov-report=xml -n auto' # comment out this line to debug a test
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply pyproject.toml:105]: It's not the messages, it's setting break points. I don't remember exactly but I think cov overwrites something from the debugger so you can't have it switch on if you want the debugger to pause at breakpoints/where there's an error

File: src/caf/distribute/gravity_model/multi_area.py [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+        class MultiCostDistribution. All zones in the cost_matrix/targets must
+        be accounted for, and should only appear in one distribution each.
+    """
+
+    # actual cost matrix
+    # adjusted cost matrix (optional) instead could be callable
+
+    # Need to update functions in base class to allow for giving multiple cost distributions
+    # Ideally calibrate and calibrate_with_perceived factors should be defined flexibly
+    # in base class
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/gravity_model/multi_area.py:157]: I don't remember writing this - I don't think it's needed anymore anyway deleting

File: src/caf/distribute/gravity_model/multi_area.py (hunk 1 above)
Comment (isaac-tfn) [reply src/caf/distribute/gravity_model/multi_area.py:236]: Yes I haven't used @overload before maybe a quick call to go over it?

File: src/caf/distribute/utils.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    infill intrazonal costs.
+    zeros_infill: The infill value for other (none diagonal) zeros in the matrix
+
+    Returns
+    -------
+    np.ndarray: The input matrix with values infilled.
+    """
+    min_row = np.min(np.ma.masked_where(cost_matrix <= 0, cost_matrix), axis=1) * diag_factor
+    np.fill_diagonal(cost_matrix, min_row)
+    cost_matrix[cost_matrix > 1e10] = zeros_infill
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/utils.py:51]: This was a bug fix - there were some diagonals being infilled with 1e20 or something. Everything was a bit of a rush just before Christmas and I forgot to get to the bottom of what was happening

File: src/caf/distribute/utils.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    cat_col: str,
+    min_col: str,
+    max_col: str,
+    ave_col: str,
+    trips_col: str,
+    tld_lookup: pd.DataFrame,
+    lookup_cat_col: str,
+    lookup_zone_col: str,
+    function_params: dict[str, float],
+) -> list[multi_area.MultiCostDistribution]:
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/utils.py:67]: I think it was originally written as an optional function for users to run, it probably should just be a class method yes

File: src/caf/distribute/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/distribute/utils.py:48]: Can this either be removed, or updated in the args / docstrings please?

File: tests/gravity_model/test_multi_area.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        assert np.array_equal(np.round(expected_infilled, 3), np.round(infilled, 3))
+
+    def test_tld_processing_length(self, distributions_sorted):
+        assert len(distributions_sorted) == 4
+
+    def test_tld_processing_type(self, distributions_sorted):
+        assert isinstance(distributions_sorted[0], gm.MultiCostDistribution)
+
+
+class TestDist:
```
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply tests/gravity_model/test_multi_area.py:187]: Please add a TODO to finish this one later

File: docs/conf.py [hunk 13]
```diff
@@ -9,20 +9,20 @@
...
+# -- Project information -----------------------------------------------------
 import sys
 import os
 from pathlib import Path
+import caf.distribute
 
 dir_path = Path(__file__).parents[1]
 source = dir_path / "src" / "caf" / "distribute"
 
 sys.path.insert(0, os.path.abspath(str(source)))
```
Comment (BenTaylor-TfN) [inline docs/conf.py:23]:
```suggestion
```

Not needed anymore?

File: docs/conf.py [hunk 14]
```diff
@@ -80,5 +80,5 @@
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ["_static"]
 
-autodoc_mock_imports = ["caf"]
+# autodoc_mock_imports = ["caf", "caf.distribute"]
```
Comment (BenTaylor-TfN) [inline docs/conf.py:83]:
```suggestion
```
No longer needed?

**Review by BenTaylor-TfN — COMMENTED** [review]
Dont forget to add to .gitblame to ignore the isort changes, just so you're not the one in the blame if something goes wrong with the isort changes!

There's still a few comments that need addressing from the previous review

File: src/caf/distribute/utils.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+# pylint: enable=import-error,wrong-import-position
+
+# # # CONSTANTS # # #
+
+# # # CLASSES # # #
+
+
+# # # FUNCTIONS # # #
+def infill_cost_matrix(
+    cost_matrix: np.ndarray, diag_factor: float = 0.5, zeros_infill: float = 0.5
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/distribute/utils.py:22]: Added as TODO

**Comment by isaac-tfn** [comment]
@BenTaylor-TfN I think I've made those changes now

**Review by BenTaylor-TfN — APPROVED** [review]
